### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by alexistr**
If replaygain has bee stored in metadata (vorbiscomment REPLAYGAIN_TRACK_GAIN) use it.
Otherwise calculate it and store it.

### CODE REVIEW

File: scripts/extract-replaygain [hunk 1]
```diff
@@ -2,22 +2,42 @@
...
 
-FFMPEG=`which ffmpeg || echo ""`
+FFMPEG=`which ffmpeg 2>/dev/null || echo ""`
+FFPROBE=`which ffprobe 2>/dev/null || echo ""`
 
 if test -z "$FFMPEG"; then
-  echo "ffmpeg must be installed"
-  exit 1
+        echo "ffmpeg must be installed"
+        exit 1
```
Comment (smimram) [inline scripts/extract-replaygain]: Why change the indentation here?
```suggestion
  echo "ffmpeg must be installed"
  exit 1
```

File: scripts/extract-replaygain [hunk 2]
```diff
@@ -2,22 +2,42 @@
...
-SED=`which gsed || echo ""`
+SED=`which gsed 2>/dev/null || echo ""`
 
 if test -z "$SED"; then
-  SED=`which sed || echo ""`
+  SED=`which sed 2>/dev/null || echo ""`
 fi
 
 if test -z "$SED"; then
   echo "sed or gsed much be installed"
```
Comment (smimram) [inline scripts/extract-replaygain]:
```suggestion
  echo "sed or gsed must be installed"
```

File: scripts/extract-replaygain [hunk 3]
```diff
@@ -2,22 +2,42 @@
...
   exit 1
 fi
 
-$FFMPEG -i "$1" -vn -filter replaygain -f null /dev/null 2>&1 | $SED -n -r 's/.* track_gain = ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'
+# check store replay gain
+RG=`$FFPROBE "$1" 2>&1 | $SED -n -r 's/.* REPLAYGAIN_TRACK_GAIN: ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'`
+
+if [ -z "$RG" ]
+then
+
```
Comment (smimram) [inline scripts/extract-replaygain]:
```suggestion
if [ -z "$RG" ]; then
```

File: scripts/extract-replaygain [hunk 4]
```diff
@@ -2,22 +2,42 @@
...
+  SED=`which sed 2>/dev/null || echo ""`
 fi
 
 if test -z "$SED"; then
   echo "sed or gsed much be installed"
   exit 1
 fi
 
-$FFMPEG -i "$1" -vn -filter replaygain -f null /dev/null 2>&1 | $SED -n -r 's/.* track_gain = ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'
+# check store replay gain
```
Comment (smimram) [inline scripts/extract-replaygain]:
```suggestion
# check for replaygain metadata
```

File: scripts/extract-replaygain [hunk 5]
```diff
@@ -2,22 +2,42 @@
...
+RG=`$FFPROBE "$1" 2>&1 | $SED -n -r 's/.* REPLAYGAIN_TRACK_GAIN: ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'`
+
+if [ -z "$RG" ]
+then
+
+ # no store replaygain : calculate it
```
Comment (smimram) [inline scripts/extract-replaygain]:
```suggestion
 # no replaygain metadata: compute it
```

File: scripts/extract-replaygain [hunk 6]
```diff
@@ -2,22 +2,42 @@
...
 
 if test -z "$FFMPEG"; then
-  echo "ffmpeg must be installed"
-  exit 1
+        echo "ffmpeg must be installed"
+        exit 1
 fi
 
-SED=`which gsed || echo ""`
+SED=`which gsed 2>/dev/null || echo ""`
```
Comment (smimram) [inline scripts/extract-replaygain]: Do we really want to hide errors here?

File: scripts/extract-replaygain (hunk 1 above)
Comment (alexistr) [reply scripts/extract-replaygain]: Sorry for the typos thank for correcting this.

File: scripts/extract-replaygain (hunk 6 above)
Comment (alexistr) [reply scripts/extract-replaygain]: Aren't these errors  handled by the `if test -z "$SED`; then" line ?

File: scripts/extract-replaygain [hunk 7]
```diff
@@ -2,22 +2,40 @@
...
+ RG=`$FFMPEG -i "$1" -vn -filter replaygain -f null /dev/null 2>&1 | $SED -n -r 's/.* track_gain = ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'`
+
+ # store it and keep file date
+ tmpdir=`mktemp -d`
+ tmpfile=$tmpdir/`basename "$1"`
+ 
+ filedate=`stat -c %y "$1"`
+ $FFMPEG -y -i "$1" -metadata REPLAYGAIN_TRACK_GAIN="$RG" -c:a copy "$tmpfile" 2> /dev/null
+
+ test -f "$tmpfile" && touch -d "$filedate" "$tmpfile" && mv "$tmpfile" "$1" && rm -r $tmpdir
```
Comment (toots) [inline scripts/extract-replaygain]: I don't think that we should write anything to the files passed to this script. These operations should be read-only by default unless explicitly stated. If you want to support this, please add an optional parameter to the replaygain protocol/caller.

File: scripts/extract-replaygain [hunk 8]
```diff
@@ -2,22 +2,40 @@
...
-$FFMPEG -i "$1" -vn -filter replaygain -f null /dev/null 2>&1 | $SED -n -r 's/.* track_gain = ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'
+# check for replaygain metadata
+RG=`$FFPROBE "$1" 2>&1 | $SED -n -r 's/.* REPLAYGAIN_TRACK_GAIN: ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'`
+
+if [ -z "$RG" ]; then
+ # no replaygain metadata: compute it
+ RG=`$FFMPEG -i "$1" -vn -filter replaygain -f null /dev/null 2>&1 | $SED -n -r 's/.* track_gain = ([-+]?[0-9]+\.[0-9]+ dB).*/\1/p'`
+
+ # store it and keep file date
+ tmpdir=`mktemp -d`
```
Comment (toots) [inline scripts/extract-replaygain]: You should add a `trap` directive to the script to make sure that this temporary folder is always removed if the script exists prematurely..

**Comment by toots** [comment]
I'm gonna take this over, make it read-only and add it to `1.4.3`. Thanks for your contribution!

**Comment by toots** [comment]
Fixed in 09f973b4 for 1.4.3

**Pull request closed without merging**